### ISSUE TITLE
Etherpad page up/page down has a substandard user experience..

### ISSUE THREAD

**Issue by JohnMcLear** [ether/etherpad-lite#4621]
# Test research findings.

1. Page down from line 0
	notepad
		Goes to one line above expected line
	winword
		Incorrect, caret in wrong place
	online word
		same as winword
	google docs
		Incorrect
	etherpad
		Incorrect, Goes to one line above expected line
	libreoffice
		Really weird, caret and focus in wrong place
	textarea

2. page down then up from caret at [0, whatever]
	notepad
		Correct
	winword
		Correct
	online word
		Correct
	google docs
		Correct
	etherpad
		Wrong, loses a line
	libreoffice
		Correct	

3.  page up / down over content that is a huge length line on 0,0
	notepad
		Correct
	winword
		Focus is correct but caret is wrong
	online word
		Same as Winword
 	google docs
		Incorrect (2 lines too far down
	etherpad
		Completely broken
	libreoffice
		Really weird, caret and focus in wrong place		

4. page down / up where notion of page break exists
	notepad
		Notion doesn't exist
	winword
		Really bizarre user experience and loses historical references
	online word
		Nearly correct (not that winword is not correct!) .. On various page downs/page up if caret lands on blank line the historical reference position is lost.
	google docs
		Loses historical references on page up after just two normal page downs with no specific line breaks etc being hit.
	etherpad
		Notion doesn't exist
	libreoffice
		Correct

5. caret location on X axis is maintained on page down/up
	notepad
		Broken, goes to character 0 on blank line
	winword
		Correct!  Very impressive, keeps indentation even after hitting blank line and visiting last character in document.
	online word
		Broken, goes to character 0 on blank line
	google docs
		Broken, goes to character 0 on blank line
	etherpad
		Broken, goes to character 0 on blank line
	libreoffice
		Broken, goes to character 0 on blank line

6. Transition / lag
	notepad
		Fast
	winword
		Smooth, Fast.
	online word
		Jerky.
	google docs
		Smooth-ish, Fast-ish.  Page identifier on scroll bar begins to land behind page location on page up/down event.
	etherpad
		Slow, Jerky.
	libreoffice
		Fast.
	
7. Page down on long text codent
	etherpad
		Broken.
	all others
		Fine.
		
# Some rules of page up / down
- [ ] Caret should always go back to position on page up after page down event (consider if this should be position after others editing or new position take into account edits [much harder to implement].
- [ ] Should always shift focus past visible content and be within a few lines of previous viewport as per spec.
- [ ] Smooth transition is a nice to have but it should be immediate.

Aria spec https://www.w3.org/TR/wai-aria-practices-1.1/ :

> Page Down (Optional): Moves focus down an author-determined number of rows, typically scrolling so the bottom row in the currently visible set of rows becomes one of the first visible rows. If focus is in the last row of the grid, focus does not move.
> 
> Page Up (Optional): Moves focus up an author-determined number of rows, typically scrolling so the top row in the currently visible set of rows becomes one of the last visible rows. If focus is in the first row of the grid, focus does not move. 

https://youtu.be/d-ZM8bozIQw is the video showing an example of each of the 7 tests.

**Comment by goatandsheep** [comment]
I didn't fully understood the tests, but here are my results for Zoho writer:

normal text, i.e. Roboto 12

* consistently 12 lines lower on page up
* page up and down move cursor to [0, whatever]
* tested on long lines, short lines, single character lines, lorem ipsum, paragraphed and non-paragraphed text and vertical columns of 0's
* transition is very smooth
* page down consistently lands on the next page

Monospaced 24px:

* page up and down works perfectly

Monospaced 12px:

* page up and down is 3 lines down

**Comment by JohnMcLear** [comment]
I recorded a video of me testing zoho and added results.  @goatandsheep https://youtu.be/wJ7EgU0awCY

I get WILDLY different results to you, so what's going on?

**Comment by JohnMcLear** [comment]
# Results of other editors.

|              | page down | page up accuracy | long line | page breaks | caret offset | transition | long text |
|--------------|-----------|------------------|-----------|-------------|--------------|------------|-----------|
| notepad      |     x     |         y        | y         |    x        |    x         |    y       |    y      |
| winword      |     x     |         y        |     x     |      x      |       y      |      y     |    y      |
| word online  |     x     |         y        |     x     |      y*2    |       x      |     x      |    y      |
| google docs  |     x     |         y        |    x      |    x        |    x         |    y*1      |    y      |
| etherpad     |     x     |         x        |    x      |    x        |    x         |    x       |    x      |
| zoho writer |     x     |         x        |    x      |    x        |    x         |    x       |    x      |
| libre office |     x     |         y        |    x      |    y        |    x         |    y       |    y      |
| textarea   |     y*4     |         y        |    y      |    x        |    x*3         |    y       |    y      |
| atom     |     x*5     |         y        |    y      |    x        |    y         |    y       |    y      |
| ckeditor     |     y    |         x*6        |    y      |    x*7        |    y         |    y       |    y      |

1 = javascript used to draw page number by scroll bar is laggy
2 = loses positioning over page breaks
3 = loses caret offset on blank line
4 = In firefox it loses line offset a little on page down but I can't consistently replicate - chrome is fine.
5 = Goes to two lines below last viewable line (should be one). 
6 = Demo at https://ckeditor.com/ckeditor-5/demo/ fails going back to correct line.
7 = Demo at https://ckeditor.com/docs/ckeditor5/latest/features/page-break.html fails to have transition between pages (loses previous line numbering)

**Comment by JohnMcLear** [comment]
Just a note from me, waiting on @webzwo0i to do a specific part but most functionality and testing is in.

**Comment by JohnMcLear** [comment]
Fixed in PR #7437 (merged) — PageDown now advances the caret by a full page of lines instead of scrolling without moving the cursor.
